### PR TITLE
feat: added retry mechanism

### PR DESCRIPTION
**Pull request by frolvanya**
Added retry logic to the `state-indexer`

### CODE REVIEW

**Comment by frolvanya** [comment]
@race-of-sloths please, include my PR in the Race

**Comment by race-of-sloths** [comment]
@frolvanya Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
Weekly streak is on the road, smart strategy! Secure your streak with another PR!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/frolvanya?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/frolvanya)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/frolvanya)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called
</details>

File: logic-state-indexer/src/lib.rs
Comment (khorolets) [inline logic-state-indexer/src/lib.rs:223]: Is there any way to emit logs to see that we've failed to save and will retry? Ideally, it should have a "recovery" log message if the retries count is greater than 1.

**Comment by khorolets** [comment]
@race-of-sloths score 1

File: logic-state-indexer/src/lib.rs
Comment (frolvanya) [reply logic-state-indexer/src/lib.rs]: Unfortunatelly `tokio_retry` didn't have a logic for tracking number of retries, so I decided to remove this dependency (technically I could introduce a mutex for this, but it's not worth it)

**Comment by frolvanya** [comment]
@khorolets @kobayurii right now we have a lot of retry logic duplication. What do you think about introducing a `fn add_retry_logic<F>(method: F, retries: usize) -> ...` utility that would wrap some custom logic and add a retry mechanism on top of it?

**Review by khorolets — APPROVED** [review]

**Pull request merged**
